File: src/clients/java/org/gridgain/client/impl/GridClientNodeImpl.java
// Copyright (C) GridGain Systems Licensed under GPLv3, http://www.gnu.org/licenses/gpl.html

/*  _________        _____ __________________        _____
 *  __  ____/___________(_)______  /__  ____/______ ____(_)_______
 *  _  / __  __  ___/__  / _  __  / _  / __  _  __ `/__  / __  __ \
 *  / /_/ /  _  /    _  /  / /_/ /  / /_/ /  / /_/ / _  /  _  / / /
 *  \____/   /_/     /_/   \_,__/   \____/   \__,_/  /_/   /_/ /_/
 */
package org.gridgain.client.impl;

import org.gridgain.client.*;

import java.net.*;
import java.util.*;
import java.util.concurrent.atomic.*;

/**
 * Client node implementation.
 *
 * @author 2012 Copyright (C) GridGain Systems
 * @version 4.0.0c.25032012
 */
public class GridClientNodeImpl implements GridClientNode {
    /** Node id. */
    private UUID nodeId;
    
    /** Internal addresses. */
    private List<String> internalAddrs = Collections.emptyList();

    /** External addresses. */
    private List<String> externalAddrs = Collections.emptyList();

    /** Port for TCP rest binary protocol. */
    private int tcpPort;

    /** Port for HTTP(S) rest binary protocol. */
    private int httpPort;

    /** Node attributes. */
    private Map<String, Object> attrs = Collections.emptyMap();

    /** Node metrics. */
    private Map<String, Object> metrics = Collections.emptyMap();

    /** Node caches. */
    private Map<String, GridClientCacheMode> caches = Collections.emptyMap();

    /** Reference to a list of addresses. */
    private AtomicReference<List<InetSocketAddress>> restAddresses = new AtomicReference<List<InetSocketAddress>>();
    
    /** {@inheritDoc} */
    @Override public UUID nodeId() {
        return nodeId;
    }

    /**
     * Node id to be set.
     *
     * @param nodeId Node id.
     */
    public void nodeId(UUID nodeId) {
        this.nodeId = nodeId;
    }

    /** {@inheritDoc} */
    @Override public List<String> internalAddresses() {
        return Collections.unmodifiableList(internalAddrs);
    }

    /**
     * Sets list of node internal addresses.
     *
     * @param internalAddrs List of addresses.
     */
    public void internalAddresses(Collection<String> internalAddrs) {
        this.internalAddrs = new ArrayList<String>(internalAddrs);
    }

    /** {@inheritDoc} */
    @Override public List<String> externalAddresses() {
        return Collections.unmodifiableList(externalAddrs);
    }

    /**
     * Sets list of node external addresses.
     *
     * @param externalAddrs List of addresses.
     */
    public void externalAddresses(Collection<String> externalAddrs) {
        this.externalAddrs = new ArrayList<String>(externalAddrs);
    }

    /** {@inheritDoc} */
    @Override public int tcpPort() {
        return tcpPort;
    }

    /**
     * @param tcpPort Sets remote port value.
     */
    public void tcpPort(int tcpPort) {
        this.tcpPort = tcpPort;
    }

    /** {@inheritDoc} */
    @Override public int httpPort() {
        return httpPort;
    }

    /**
     * Sets remote http port value.
     *
     * @param httpPort Http(s) port value.
     */
    public void httpPort(int httpPort) {
        this.httpPort = httpPort;
    }

    /** {@inheritDoc} */
    @Override public Map<String, Object> attributes() {
        return Collections.unmodifiableMap(attrs);
    }

    /**
     * Sets node attributes.
     *
     * @param attrs Node attributes.
     */
    public void attributes(Map<String, Object> attrs) {
        this.attrs = attrs;
    }

    /** {@inheritDoc} */
    @SuppressWarnings("unchecked")
    @Override public <T> T attribute(String name) {
        return (T)attrs.get(name);
    }

    /** {@inheritDoc} */
    @Override public Map<String, Object> metrics() {
        return metrics;
    }

    /**
     * Sets node metrics.
     *
     * @param metrics Metrics.
     */
    public void metrics(Map<String, Object> metrics) {
        this.metrics = metrics;
    }

    /** {@inheritDoc} */
    @Override public Map<String, GridClientCacheMode> caches() {
        return caches;
    }

    /**
     * Sets caches available on remote node.
     *
     * @param caches Cache map.
     */
    public void caches(Map<String, GridClientCacheMode> caches) {
        this.caches = caches;
    }

    /**
     * Gets list of all addresses available for connection for tcp rest binary protocol.
     *
     * @param proto Protocol type.
     * @return List of socket addresses.
     */
    @Override public List<InetSocketAddress> availableAddresses(GridClientProtocol proto) {
        List<InetSocketAddress> res = restAddresses.get();
        
        if (res == null) {
            res = new ArrayList<InetSocketAddress>(internalAddrs.size() + externalAddrs.size());

            int port = proto == GridClientProtocol.TCP ? tcpPort : httpPort;

            if (port != 0) {
                for (String internalAddr : internalAddrs)
                    res.add(new InetSocketAddress(internalAddr, port));

                for (String internalAddr : internalAddrs)
                    res.add(new InetSocketAddress(internalAddr, port));
            }
            
            restAddresses.compareAndSet(null, res);
            
            res = restAddresses.get();
        }
        
        assert res != null;
        
        return res;
    }

    /** {@inheritDoc} */
    @Override public boolean equals(Object o) {
        if (this == o) return true;

        if (!(o instanceof GridClientNodeImpl)) return false;

        GridClientNodeImpl that = (GridClientNodeImpl)o;

        return nodeId.equals(that.nodeId);
    }

    /** {@inheritDoc} */
    @Override public int hashCode() {
        return nodeId.hashCode();
    }

    @Override
    public String toString() {
        final StringBuilder sb = new StringBuilder("GridClientNodeImpl [nodeId=");

        sb.append(nodeId);
        sb.append(", internalAddrs=").append(internalAddrs);
        sb.append(", externalAddrs=").append(externalAddrs);
        sb.append(", binaryPort=").append(tcpPort);
        sb.append(", httpPort=").append(httpPort);
        sb.append(']');

        return sb.toString();
    }
}

File: src/clients/dotnet/client/impl/GridClientNodeImpl.cs
// Copyright (C) GridGain Systems Licensed under GPLv3, http://www.gnu.org/licenses/gpl.html

/*  _________        _____ __________________        _____
 *  __  ____/___________(_)______  /__  ____/______ ____(_)_______
 *  _  / __  __  ___/__  / _  __  / _  / __  _  __ `/__  / __  __ \
 *  / /_/ /  _  /    _  /  / /_/ /  / /_/ /  / /_/ / _  /  _  / / /
 *  \____/   /_/     /_/   \_,__/   \____/   \__,_/  /_/   /_/ /_/
 */

namespace GridGain.Client.Impl {
    using System;
    using System.Collections.Generic;
    using System.Threading;
    using System.Text;
    using System.Net;
    using GridGain.Client;
    using GridGain.Client.Util;
    using GridGain.Client.Hasher;

    using A = GridGain.Client.Util.GridClientArgumentCheck;
    using Dbg = System.Diagnostics.Debug;

    /** <summary>Client node implementation.</summary> */
    internal class GridClientNodeImpl : IGridClientNode, IGridClientConsistentHashObject {
        /** <summary>Reference to a list of addresses.</summary> */
        private readonly HashSet<IPEndPoint> restAddresses = new HashSet<IPEndPoint>();

        /**
         * <summary>
         * Constructs grid client node.</summary>
         *
         * <param name="nodeId">Node ID.</param>
         */
        public GridClientNodeImpl(Guid nodeId) {
            Id = nodeId;
            InternalAddresses = new List<String>();
            ExternalAddresses = new List<String>();
            Attributes = new Dictionary<String, Object>();
            Metrics = null;
            Caches = new GridClientNullDictionary<String, GridClientCacheMode>();
        }

        /** <summary>Node id.</summary> */
        public Guid Id {
            get;
            private set;
        }

        /** <summary>List of node internal addresses.</summary> */
        public IList<String> InternalAddresses {
            get;
            private set;
        }

        /** <summary>List of node external addresses.</summary> */
        public IList<String> ExternalAddresses {
            get;
            private set;
        }

        /** <summary>Tcp remote port value.</summary> */
        public int TcpPort {
            get;
            set;
        }

        /** <summary>Http(s) port value.</summary> */
        public int HttpPort {
            get;
            set;
        }

        /** <summary>Node attributes.</summary> */
        public IDictionary<String, Object> Attributes {
            get;
            private set;
        }

        /** <inheritdoc /> */
        public T Attribute<T>(String name) {
            return Attribute<T>(name, default(T));
        }

        /** <inheritdoc /> */
        public T Attribute<T>(String name, T def) {
            Object result;

            return Attributes.TryGetValue(name, out result) && (result is T) ? (T)result : def;
        }


        /** <summary>Node metrics.</summary> */
        public IGridClientNodeMetrics Metrics {
            get;
            set;
        }

        /** Caches available on remote node.*/
        public IDictionary<String, GridClientCacheMode> Caches {
            get;
            private set;
        }

        /**
         * <summary>
         * Gets list of all addresses available for connection for tcp rest binary protocol.</summary>
         *
         * <param name="proto">Protocol type.</param>
         * <returns>List of socket addresses.</returns>
         */
        public IList<IPEndPoint> AvailableAddresses(GridClientProtocol proto) {
            lock (restAddresses) {
                if (restAddresses.Count == 0) {
                    int port = proto == GridClientProtocol.Tcp ? TcpPort : HttpPort;

                    if (port != 0) {
                        foreach (String addr in InternalAddresses)
                            restAddresses.Add(new IPEndPoint(Dns.GetHostAddresses(addr)[0], port));

                        foreach (String addr in ExternalAddresses)
                            restAddresses.Add(new IPEndPoint(Dns.GetHostAddresses(addr)[0], port));
                    }
                }
            }

            Dbg.Assert(restAddresses.Count > 0, "at least one rest address is defined");

            return new List<IPEndPoint>(restAddresses);
        }

        /** <inheritdoc /> */
        override public bool Equals(Object obj) {
            if (this == obj)
                return true;

            GridClientNodeImpl that = obj as GridClientNodeImpl;

            return that != null && Id.Equals(that.Id);
        }

        /** <inheritdoc /> */
        override public int GetHashCode() {
            return Id.GetHashCode();
        }

        /** <inheritdoc /> */
        override public String ToString() {
            StringBuilder sb = new StringBuilder("GridClientNodeImpl");

            sb.AppendFormat(" [NodeId={0}", Id);
            sb.AppendFormat(", InternalAddresses={0}", InternalAddresses);
            sb.AppendFormat(", ExternalAddresses={0}", ExternalAddresses);
            sb.AppendFormat(", TcpPort={0}", TcpPort);
            sb.AppendFormat(", HttpPort={0}", HttpPort);
            sb.Append(']');

            return sb.ToString();
        }
    }
}
